src/Tasks/Unzip.cs

```diff
@@ -0,0 +1,206 @@
+
+                return false;
+            }
+
+            BuildEngine3.Yield();
+
+            try
+            {
+                foreach (ITaskItem sourceFile in SourceFiles.TakeWhile(i => !_cancellationToken.IsCancellationRequested))
+                {
+                    if (!File.Exists(sourceFile.ItemSpec))
+                    {
+                        Log.LogErrorFromResources("Unzip.ErrorFileDoesNotExist", sourceFile.ItemSpec);
+                        continue;
+                    }
+
+                    try
+                    {
+                        using (FileStream stream = new FileStream(sourceFile.ItemSpec, FileMode.Open, FileAccess.Read, FileShare.Read, bufferSize: 0x1000, useAsync: false))
+                        {
+                            using (ZipArchive zipArchive = new ZipArchive(stream, ZipArchiveMode.Read, leaveOpen: false))
+                            {
+                                try
+                                {
+                                    Extract(zipArchive, destinationDirectory);
+                                }
+                                catch (Exception e)
+                                {
+                                    // Unhandled exception in Extract() is a bug!
+                                    Log.LogErrorFromException(e, showStackTrace: true);
+                                    return false;
+                                }
+                            }
+                        }
+                    }
+                    catch (OperationCanceledException)
+                    {
+                        break;
+                    }
+                    catch (Exception e)
+                    {
+                        // Should only be thrown if the archive could not be opened (Access denied, corrupt file, etc)
+                        Log.LogErrorFromResources("Unzip.ErrorCouldNotOpenFile", sourceFile.ItemSpec, e.Message);
+                    }
+                }
+            }
+            finally
+            {
+                BuildEngine3.Reacquire();
+            }
+
+            return !_cancellationToken.IsCancellationRequested && !Log.HasLoggedErrors;
+        }
+
+        /// <summary>
+        /// Extracts all files to the specified directory.
+        /// </summary>
+        /// <param name="sourceArchive">The <see cref="ZipArchive"/> containing the files to extract.</param>
+        /// <param name="destinationDirectory">The <see cref="DirectoryInfo"/> to extract files to.</param>
+        private void Extract(ZipArchive sourceArchive, DirectoryInfo destinationDirectory)
+        {
+            foreach (ZipArchiveEntry zipArchiveEntry in sourceArchive.Entries.TakeWhile(i => !_cancellationToken.IsCancellationRequested))
+            {
+                FileInfo destinationPath = new FileInfo(Path.Combine(destinationDirectory.FullName, zipArchiveEntry.FullName));
```

Comment:
Sorry to comment after you've just merged, but would you consider normalizing directory separators? One of the most common errors I've encountered is that zips created on Windows using `\` as a directory separator will not extract correctly on Linux/macOS. Instead `\` becomes part of the filename.
![image](https://user-images.githubusercontent.com/2696087/40389159-71632bd8-5dc6-11e8-9da8-f58484b39d3c.png)

This common error is something we worked around in the unzip task used in the .NET Core build system: see https://github.com/dotnet/arcade/blob/0d117c0b3649565a6d9aacf9f435e29ab67c3c0d/src/Microsoft.DotNet.Build.Tasks.IO/src/UnzipArchive.cs#L64-L76